network/discovery.go

```diff
@@ -218,54 +250,125 @@ func (d *discovery) getStream(peerID peer.ID) (interface{}, error) {
-
-	for indx, node := range resp.Nodes {
-		info, err := StringToAddrInfo(node)
-		if err != nil {
-			return nil, err
-		}
-
-		addrInfo[indx] = info
-	}
-
-	return addrInfo, nil
+	return resp.Nodes, nil
 }
 
 func (d *discovery) run() {
+	peerDiscoveryTicker := time.NewTicker(peerDiscoveryInterval)
+	bootnodeDiscoveryTicker := time.NewTicker(bootnodeDiscoveryInterval)
+
+	defer func() {
+		peerDiscoveryTicker.Stop()
+		bootnodeDiscoveryTicker.Stop()
+	}()
+
 	for {
 		select {
-		case <-time.After(5 * time.Second):
-		case <-d.notifyCh:
 		case <-d.closeCh:
 			return
+		case <-peerDiscoveryTicker.C:
+			go d.handleDiscovery()
+		case <-bootnodeDiscoveryTicker.C:
+			go d.bootnodeDiscovery()
 		}
-		d.handleDiscovery()
 	}
 }
 
 func (d *discovery) handleDiscovery() {
 	// take a random peer and find peers
-	if target := d.peers.getRandomPeer(); target != nil {
-		if err := d.attemptToFindPeers(target.id); err != nil {
-			d.srv.logger.Error("failed to dial peer", "peer", target.id, "err", err)
+	if d.srv.availableOutboundConns() > 0 {
+		if target := d.peers.getRandomPeer(); target != nil {
+			if err := d.attemptToFindPeers(target.id); err != nil {
+				d.srv.logger.Error("Failed to find new peers", "peer", target.id, "err", err)
+			}
 		}
 	}
 }
 
+func (d *discovery) getBootNodeConnCount() int32 {
+	return atomic.LoadInt32(&d.bootnodeConnCount)
+}
+
+// bootnodeDiscovery queries a random bootnode for new peers and adds them to the routing table
+func (d *discovery) bootnodeDiscovery() {
+	if d.srv.availableOutboundConns() <= 0 {
+		return
+	}
+
+	// get a random bootnode which is not connected
+	bootnode := d.srv.getBootNode()
+	if bootnode == nil {
+		return
+	}
+
+	// isTemporaryDial maintains the dial status
+	var isTemporaryDial bool
+
+	// if one or more bootnode is connected the dial status is temporary
+	if d.getBootNodeConnCount() >= 1 {
+		isTemporaryDial = true
+	}
+
+	if isTemporaryDial {
+		if _, loaded := d.srv.temporaryDials.LoadOrStore(bootnode.ID, true); loaded {
+			return
+		}
+	}
+
+	defer func() {
+		if isTemporaryDial {
+			d.srv.temporaryDials.Delete(bootnode.ID)
+		}
+	}()
+
+	if len(d.srv.host.Peerstore().Addrs(bootnode.ID)) == 0 {
+		d.srv.host.Peerstore().AddAddr(bootnode.ID, bootnode.Addrs[0], peerstore.AddressTTL)
+	}
+
+	stream, err := d.srv.NewProtoStream(discProto, bootnode.ID)
+	if err != nil {
+		d.srv.logger.Error("Failed to open new stream", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	clt := proto.NewDiscoveryClient(stream.(*rawGrpc.ClientConn))
+
+	resp, err := clt.FindPeers(context.Background(), &proto.FindPeersReq{Count: defaultPeerReqCount})
+	if err != nil {
+		d.srv.logger.Error("Find peers call failed", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	if err := stream.(*rawGrpc.ClientConn).Close(); err != nil {
+		d.srv.logger.Error("Error closing grpc stream", "peer", bootnode.ID, "err", err)
+
+		return
+	}
+
+	if isTemporaryDial {
+		d.srv.Disconnect(bootnode.ID, "Thank you")
+	}
```

Comment:
You added the same code inside defer above

Reply:
removed it, commit-[link](https://github.com/0xPolygon/polygon-edge/pull/343/commits/d4489ac2fdfeb274d783226a8aed1da5d13b99ea)